### PR TITLE
REM-1130: Implement obligatory payment system

### PR DESCRIPTION
**Pull request by Art17**
### Jira references
- Ticket (task) — https://remmeio.atlassian.net/browse/REM-1130

### CODE REVIEW

File: remme/tp/obligatory_payment.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+LOGGER = logging.getLogger(__name__)
+
+FAMILY_NAME = 'obligatory_payment'
+FAMILY_VERSIONS = ['0.1']
+
+
+def withdraw_obligatory_payment(node_account, obligatory_payment):
+    if node_account.balance >= obligatory_payment:
+        node_account.balance -= obligatory_payment
+    elif node_account.reputation.unfrozen >= obligatory_payment:
```
Comment (dmytrostriletskyi) [inline remme/tp/obligatory_payment.py]: Put a blank line between elifs. Will be more readable.

```
if ...
    ....

elif ...
    ....

elif:
   ...
```

File: remme/tp/obligatory_payment.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        node_account.reputation.unfrozen -= obligatory_payment
+    elif node_account.reputation.frozen >= obligatory_payment:
+        node_account.reputation.frozen -= obligatory_payment
+    else:
```
Comment (dmytrostriletskyi) [inline remme/tp/obligatory_payment.py]: You do not need `else`, just raise.

File: remme/tp/obligatory_payment.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+        committee_pub_keys, obligatory_payment, committee_size = get_obligatory_payment_parameters(context)
+
+        address_to_node_account_dict = {
+            node_account_address: node_account,
+        }
+
+        for pub_key in committee_pub_keys:
+            if pub_key == node_account_public_key:
+                continue
```
Comment (dmytrostriletskyi) [inline remme/tp/obligatory_payment.py]: Migrate to:

```
committee_to_charge_obligatory_payment = committee_pub_keys.remove(node_account_public_key)
```

For me it is more obvious why you pass node public key.

File: remme/tp/obligatory_payment.py (hunk 1 above)
Comment (Art17) [reply remme/tp/obligatory_payment.py]: Done.

File: remme/tp/obligatory_payment.py (hunk 2 above)
Comment (Art17) [reply remme/tp/obligatory_payment.py]: else needed.

File: remme/tp/obligatory_payment.py (hunk 3 above)
Comment (Art17) [reply remme/tp/obligatory_payment.py]: Done.

File: remme/tp/obligatory_payment.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+        if node_account is None:
+            raise InvalidTransaction('Invalid context or address.')
+
+        committee_pub_keys, obligatory_payment, committee_size = get_obligatory_payment_parameters(context)
+
+        address_to_node_account_dict = {
+            node_account_address: node_account,
+        }
+        committee_pub_keys.remove(node_account_public_key)  # committee to charge obligatory payment
```
Comment (eugene-babichenko) [inline remme/tp/obligatory_payment.py]: Why do that?

File: remme/tp/obligatory_payment.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    elif node_account.reputation.unfrozen >= obligatory_payment:
+        node_account.reputation.unfrozen -= obligatory_payment
+
+    elif node_account.reputation.frozen >= obligatory_payment:
+        node_account.reputation.frozen -= obligatory_payment
+
+    else:
+        raise InvalidTransaction("Malformed committee. A node doesn't have enough tokens to pay obligatory payment.")
+
+def get_obligatory_payment_parameters(context):
```
Comment (eugene-babichenko) [inline remme/tp/obligatory_payment.py]: This permissioning setup must be changed to what is done in #270.

File: scripts/node/validator.sh [hunk 6]
```diff
@@ -61,8 +61,8 @@ if [ "$REMME_START_MODE" = "genesis" ]; then
     echo "Writing batch injector settings..."
     sawset proposal create \
         -k /etc/sawtooth/keys/validator.priv \
-        sawtooth.validator.batch_injectors=block_info \
-        "sawtooth.validator.block_validation_rules=NofX:1,block_info;XatY:block_info,0;local:0" \
+        "sawtooth.validator.batch_injectors=block_info;obligatory_payment" \
```
Comment (eugene-babichenko) [inline scripts/node/validator.sh]: 1. comma should be used, not semicolon
2. we will create our own BatchInjector for all transactions that should be enforced

File: remme/tp/obligatory_payment.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        return {
+            ObligatoryPaymentMethod.PAY_OBLIGATORY_PAYMENT: {
+                PB_CLASS: ObligatoryPaymentPayload,
+                PROCESSOR: self._pay_obligatory_payment,
+                VALIDATOR: ObligatoryPaymentPayloadForm,
+            },
+        }
+
+    def _pay_obligatory_payment(self, context, node_account_public_key, obligatory_payment_payload):
+        node_account_address = self.make_address_from_data(node_account_public_key)
```
Comment (eugene-babichenko) [inline remme/tp/obligatory_payment.py]: This address should be in the `NodeAccount` transaction family.

File: remme/tp/obligatory_payment.py (hunk 4 above)
Comment (DeV1doR) [reply remme/tp/obligatory_payment.py]: Could be a possible  ValueError

**Comment by eugene-babichenko** [comment]
Was rebased to `dev-consensus` in #272, @dmytrostriletskyi I propose to close this one.

**Pull request closed without merging**